File: app/src/main/java/sgtmelon/scriptum/office/utils/TimeUtils.java
package sgtmelon.scriptum.office.utils;

import android.content.Context;
import android.text.format.DateUtils;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import androidx.annotation.NonNull;
import sgtmelon.scriptum.R;

/**
 * Класс для работы со временем, а так же @Singleton для ***
 */
public final class TimeUtils {

    @NonNull
    private static SimpleDateFormat getDateFormat(@NonNull Context context) {
        return new SimpleDateFormat(context.getString(R.string.date_app_format), Locale.getDefault());
    }

    /**
     * @return - Текущее время в нужном формате
     */
    public static String getTime(@NonNull Context context) {
        return getDateFormat(context).format(Calendar.getInstance().getTime());
    }

    /**
     * @param date - Время создания/изменения заметки
     * @return - Время и дата в приятном виде
     */
    static String format(@NonNull Context context, @NonNull String date) {
        final DateFormat formatOld = getDateFormat(context);

        try {
            final Calendar calendar = Calendar.getInstance();
            calendar.setTime(formatOld.parse(date));

            final DateFormat formatNew = new SimpleDateFormat(
                    DateUtils.isToday(calendar.getTimeInMillis())
                            ? context.getString(R.string.date_note_today_format)
                            : context.getString(R.string.date_note_yesterday_format),
                    Locale.getDefault()
            );

            return formatNew.format(formatOld.parse(date));
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/sgtmelon/scriptum/office/utils/TimeUtils.kt
package sgtmelon.scriptum.office.utils

import android.content.Context
import android.text.format.DateUtils
import sgtmelon.scriptum.R
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*


/**
 * Класс для работы со временем, а так же @Singleton для ***
 */
object TimeUtils {

    private fun Context.getDateFormat(): SimpleDateFormat =
            SimpleDateFormat(getString(R.string.date_app_format), Locale.getDefault())

    private fun Locale.useAmPm(): Boolean =
            (DateFormat.getTimeInstance(DateFormat.FULL, this)).let {
                return it is SimpleDateFormat && it.toPattern().contains("a")
            }

    /**
     * Текущее время в нужном формате
     */
    fun Context.getTime(): String = getDateFormat().format(Calendar.getInstance().time)

    /**
     * Форматирует прошедшее время [date] в приятный вид
     */
    fun formatPast(context: Context, date: String): String? = try {
        val calendar = Calendar.getInstance().apply { time = context.getDateFormat().parse(date) }
        val locale = Locale.getDefault()

        SimpleDateFormat(when {
            calendar.isToday() -> if (locale.useAmPm()) context.getString(R.string.format_time_am) else context.getString(R.string.format_time)
            calendar.isThisYear() -> context.getString(R.string.format_date_medium)
            else -> context.getString(R.string.format_date_short)
        }, locale).format(calendar.time)
    } catch (e: Throwable) {
        null
    }

    private fun Calendar.isToday() = DateUtils.isToday(timeInMillis)

    private fun Calendar.isThisYear() = get(Calendar.YEAR) == Calendar.getInstance().get(Calendar.YEAR)

    fun formatFuture(context: Context, date: String): String? = try {
        val calendar = Calendar.getInstance().apply { time = context.getDateFormat().parse(date) }

        DateUtils.getRelativeDateTimeString(context, calendar.timeInMillis,
                DateUtils.DAY_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0
        ).toString()
    } catch (e: Throwable) {
        null
    }

}